Talk about concurrency in H3 vs. H2 (#3114)
quicwg/base-drafts: draft-ietf-quic-http.md

space is significantly larger such that it is likely that other limits in QUIC
are reached first, such as the limit on the connection flow control window.

> QUIC manages stream concurrency on behalf of HTTP/3.  QUIC considers a stream

Reviewer: I feel this "on behalf of HTTP/3" a bit confusing, because my mental model is that it is generally a responsibility of the transport to manage streams, and that HTTP/2 was the exception.

Maybe something like "In contrast to HTTP/2, stream concurrency in HTTP/3 is managed by QUIC."

closed when all sent data and all acknowledgements for received data have been

> acknowledged by the peer.  HTTP/2 considers a stream closed when the frame

Reviewer: I think this sentence is incorrect.

I think it should be: "QUIC considers a stream closed when all data have been received and all the sent data have been acknowledged."

Reviewer 2: Or a reset has been sent and acknowledged.

containing the END_STREAM bit has been committed to the transport. As a result,
the stream for an equivalent exchange will remain "active" for at least one

> additional round trip.  Implementations of HTTP/3 might choose to permit a

Reviewer: How about "will typically remain "active" for one additional roundtrip"?

The lifetime of a stream would be indifferent from HTTP/2 in the pathological case where the request-side of the stream is closed after the response-side is being closed.

space is significantly larger such that it is likely that other limits in QUIC
are reached first, such as the limit on the connection flow control window.

In contrast to HTTP/2, stream concurrency in HTTP/3 is managed by QUIC.  QUIC
considers a stream closed when all data has been received and sent data has been
acknowledged by the peer.  HTTP/2 considers a stream closed when the frame
containing the END_STREAM bit has been committed to the transport. As a result,
the stream for an equivalent exchange will typically remain "active" for one
additional round trip.  HTTP/3 servers might choose to permit a larger number of
concurrent client-initiated bidirectional streams to achieve equivalent

> concurrency than were permitted in HTTP/2, depending on the expected usage

Reviewer 3: "permit a larger number...than were permitted" makes sense, but there are so many words between them that it's kind of hard to parse.

> concurrency than were permitted in HTTP/2, depending on the expected usage

Reviewer 2: ```suggestion
concurrency to in HTTP/2, depending on the expected usage
```

patterns.

Due to the presence of other unidirectional stream types, HTTP/3 does not rely
exclusively on the number of concurrent unidirectional streams to control the

> number of concurrent pushes received by a client.  Instead, HTTP/3 clients use

Reviewer 4: ```suggestion
number of concurrent in-flight pushes.  Instead, HTTP/3 clients use
```

Avoid the rabbit whole of received by client or sent by server.
